File: common/src/main/scripts/lib/micropython/oc2/ports.py
import ffi
import os

PORTS_PATH = "/sys/class/virtio-ports"


def find(name):
    try:
        entries = [entry[0] for entry in os.ilistdir(PORTS_PATH)]
    except OSError:
        return None

    for entry in entries:
        try:
            with open("%s/%s/name" % (PORTS_PATH, entry)) as f:
                if f.read().strip() == name:
                    return "/dev/%s" % entry
        except Exception:
            pass
    return None


_libc = ffi.open(None)
_fcntl = _libc.func("i", "fcntl", "iii")
_F_GETFL = 3
_F_SETFL = 4
_F_GETFD = 1
_F_SETFD = 2
_O_NONBLOCK = 0o4000
_FD_CLOEXEC = 1


def set_nonblocking(fd):
    flags = _fcntl(fd, _F_GETFL, 0)
    if flags < 0 or _fcntl(fd, _F_SETFL, flags | _O_NONBLOCK) < 0:
        raise OSError("could not set O_NONBLOCK on the descriptor")


def set_cloexec(fd):
    flags = _fcntl(fd, _F_GETFD, 0)
    if flags >= 0:
        _fcntl(fd, _F_SETFD, flags | _FD_CLOEXEC)

File: common/src/main/scripts/lib/lua/oc2/ports.lua
local ports = {}

local ports_path = "/sys/class/virtio-ports"

local function readPortName(entry)
  local file = io.open(ports_path .. "/" .. entry .. "/name", "r")
  if not file then
    return nil
  end
  local value = file:read("l")
  file:close()
  return value
end

function ports.find(name)
  local ok, dirent = pcall(require, "posix.dirent")
  if not ok then
    return nil
  end

  local listed, entries = pcall(dirent.dir, ports_path)
  if not listed or not entries then
    return nil
  end

  for _, entry in ipairs(entries) do
    if entry ~= "." and entry ~= ".." then
      if readPortName(entry) == name then
        return "/dev/" .. entry
      end
    end
  end
end

return ports
